File: common-lib/src/main/java/com/popcorn/common/kafka/config/KafkaConfig.java
package com.popcorn.common.kafka.config;

import lombok.extern.slf4j.Slf4j;

import org.apache.kafka.clients.admin.NewTopic;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.common.serialization.StringSerializer;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.kafka.annotation.EnableKafka;
import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
import org.springframework.kafka.config.TopicBuilder;
import org.springframework.kafka.core.*;
import org.springframework.kafka.listener.DeadLetterPublishingRecoverer;
import org.springframework.kafka.listener.DefaultErrorHandler;
import org.springframework.kafka.support.ExponentialBackOffWithMaxRetries;
import org.springframework.kafka.support.serializer.JsonDeserializer;
import org.apache.kafka.common.config.TopicConfig;
import org.apache.kafka.common.serialization.StringDeserializer;
import org.apache.kafka.common.serialization.StringSerializer;
import org.apache.kafka.common.TopicPartition;

import java.util.HashMap;
import java.util.Map;

/**
 * Kafka 연결 설정을 관리하는 관리파일.
 *
 * @author : jonghoon
 * @fileName : KafkaConfig
 * @since : 25. 1. 9.
 */
@Slf4j
@EnableKafka
@Configuration
public class KafkaConfig {

    // Kafka 연결 서버 주소
    @Value("${spring.kafka.bootstrap-servers}")
    private String bootstrapServers;

    @Value("${spring.kafka.consumer.group-id}")
    private String groupId;

    /**
     * Producer 설정을 위한 Factory Bean
     * - 메시지 생산자의 직렬화 설정 및 서버 연결 설정을 담당
     *
     * @return ProducerFactory<String, Object>
     */
    @Bean
    public ProducerFactory<String, Object> producerFactory() {
        Map<String, Object> props = new HashMap<>();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, org.apache.kafka.common.serialization.StringSerializer.class);
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, org.springframework.kafka.support.serializer.JsonSerializer.class);
        return new DefaultKafkaProducerFactory<>(props);
    }

    @Bean
    public KafkaTemplate<String, Object> kafkaTemplate() {
        return new KafkaTemplate<>(producerFactory());
    }

     /**
     * Consumer 설정을 위한 Factory Bean
     * 메시지 소비자의 역직렬화 설정 및 그룹 ID, 오프셋 설정을 담당
     *
     * @return ConsumerFactory<String, String>
     */
    @Bean
    public ConsumerFactory<String, String> consumerFactory() {
        Map<String, Object> props = new HashMap<>();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        props.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
        //props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
        // at-least-once 전제 -> 수동 커밋
        //props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
        // Key 역직렬화
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        //props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, JsonDeserializer.class);
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        return new DefaultKafkaConsumerFactory<>(props);
    }

    /**
     * Kafka Listener 컨테이너 Factory Bean
     * 어노테이션 @KafkaListener 메서드들을 위한 컨테이너 설정
     *
     * @return ConcurrentKafkaListenerContainerFactory<String, String>
     */
    @Bean
    public ConcurrentKafkaListenerContainerFactory<String, String> kafkaListenerContainerFactory() {
        ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
        factory.setConsumerFactory(consumerFactory());
        // 실패 시 Retry 후 DLT로
        //factory.setCommonErrorHandler(errorHandler);
        return factory;
    }

    /**
     * 실패 처리 정책:
     * - 재시도(지수 백오프)
     * - 최대 초과 시 <원본토픽>.DLT 로 전송
     *
     * 예) store-requests.DLT
     */
    /*@Bean
    public DefaultErrorHandler errorHandler(KafkaTemplate<String, String> template) {
        DeadLetterPublishingRecoverer recoverer = new DeadLetterPublishingRecoverer(
                template,
                (record, ex) -> new TopicPartition(record.topic() + ".DLT", record.partition())
        );

        ExponentialBackOffWithMaxRetries backOff = new ExponentialBackOffWithMaxRetries(6);
        backOff.setInitialInterval(1_000L);
        backOff.setMultiplier(2.0);
        backOff.setMaxInterval(30_000L);

        DefaultErrorHandler handler = new DefaultErrorHandler(recoverer, backOff);

        // 재시도 의미 없는 예외는 즉시 DLT로 보내고 싶으면 여기에 추가
        // handler.addNotRetryableExceptions(IllegalArgumentException.class);

        return handler;
    }*/

    @Bean
    public NewTopic orderEventsTopic() {
        return TopicBuilder.name("order-events")
                .partitions(3)
                .replicas(1)
                .config(TopicConfig.RETENTION_MS_CONFIG, "86400000") // 1일
                .config(TopicConfig.COMPRESSION_TYPE_CONFIG, "snappy")
                .build();
    }

    @Bean
    public NewTopic paymentEventsTopic() {
        return TopicBuilder.name("payment-events")
                .partitions(3)
                .replicas(1)
                .config(TopicConfig.RETENTION_MS_CONFIG, "86400000")
                .build();
    }

}

File: coupon/src/main/kotlin/com/popcorn/coupon/config/KafkaConfig.kt
package com.popcorn.coupon.config

import org.apache.kafka.clients.admin.NewTopic
import org.apache.kafka.common.config.TopicConfig
import org.slf4j.LoggerFactory
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.kafka.config.TopicBuilder

/**
 * 쿠폰 서비스 전용 DLQ/Retry 토픽 설정
 * - common-lib의 KafkaConfig를 사용하고, 추가 토픽만 정의
 */
@Configuration
@ConditionalOnProperty(value = ["kafka.enabled"], havingValue = "true")
class CouponKafkaConfig {

    private val log = LoggerFactory.getLogger(CouponKafkaConfig::class.java)

    // === 쿠폰 전용 DLQ/Retry 토픽들 ===

    @Bean
    fun couponEventsRetryTopic(): NewTopic {
        return TopicBuilder.name("coupon-events-retry")
            .partitions(6)
            .replicas(1)
            .config(TopicConfig.RETENTION_MS_CONFIG, "604800000") // 7일 보관
            .config(TopicConfig.COMPRESSION_TYPE_CONFIG, "snappy")
            .build()
    }

    @Bean
    fun couponEventsDlqTopic(): NewTopic {
        return TopicBuilder.name("coupon-events-dlq")
            .partitions(6)
            .replicas(1)
            .config(TopicConfig.RETENTION_MS_CONFIG, "2592000000") // 30일 보관
            .config(TopicConfig.COMPRESSION_TYPE_CONFIG, "snappy")
            .build()
    }

    @Bean
    fun couponRequestsRetryTopic(): NewTopic {
        return TopicBuilder.name("coupon-requests-retry")
            .partitions(6)
            .replicas(1)
            .config(TopicConfig.RETENTION_MS_CONFIG, "604800000") // 7일 보관
            .config(TopicConfig.COMPRESSION_TYPE_CONFIG, "snappy")
            .build()
    }

    @Bean
    fun couponRequestsDlqTopic(): NewTopic {
        return TopicBuilder.name("coupon-requests-dlq")
            .partitions(6)
            .replicas(1)
            .config(TopicConfig.RETENTION_MS_CONFIG, "2592000000") // 30일 보관
            .config(TopicConfig.COMPRESSION_TYPE_CONFIG, "snappy")
            .build()
    }

    @Bean
    fun couponAnalyticsEventsTopic(): NewTopic {
        return TopicBuilder.name("coupon-analytics-events")
            .partitions(12)
            .replicas(1)
            .config(TopicConfig.RETENTION_MS_CONFIG, "1209600000") // 14일 보관 (분석 데이터)
            .config(TopicConfig.COMPRESSION_TYPE_CONFIG, "snappy")
            .build()
    }
}